Financialcircuitgrant

fosgate29 opened the pull request:
# Description

Add ZKP Financial Circuit as an example. 

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

Financial zkp circuits can be used as an example and be reused by other developers

## How Has This Been Tested

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate)

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

humbitious reviewed (dismissed):
Reviewed the Readme. Looks good.

myrade commented on examples/financialcircuit/testcases.go:
```diff
@@ -0,0 +1,144 @@
+
+func getQuotesValue(bond *Bond, quoteCpt1, quoteCpt2, quoteCpt3, message string) TestCase {
+
+	_quoteCpt1, err := decimal.NewFromString(quoteCpt1)
```
err values on lines 96-97 are not checked

fosgate29 replied:
good. I will review err values and checks

myrade reviewed (commented):
There are a number of unchecked and unassigned `err` returns.

myrade commented:
I tested this circuit with the most recent gnark commit for curiosity's sake since performance was mentioned in the code and docs, and I found it has vastly improved runtime. I pushed a commit here if anyone would like to test as it required a good few edits due to new syntax: https://github.com/myrade/baseline/commit/e1ded506b50f72975ee5015fb7d7b628e22c9435

Results on my system are 5.4 times better speed on circuit setup (1.2s vs 6.5s) and 3.6 times better speed on the entire test suite (4.8s vs 17.3s).

fosgate29 commented on examples/financialcircuit/bond.go:
```diff
@@ -0,0 +1,137 @@
+	checkZeroTemp := cs.Or(checkZeroCpt1, checkZeroCpt2)
+	checkZeroResult := cs.Or(checkZeroTemp, checkZeroCpt3)
+
+	cs.AssertIsEqual(checkZeroResult, 0)
```
good catch. I will fix it.

fosgate29 commented:
I will work on suggestions made by @myrade and update the PR

GoldenBit0 commented:
@fosgate29 Thanks! Let me know once you're done and we'll get the PR reviewed by other Maintainers.

fosgate29 commented:
@myrade Hi. Thanks for the review. Code is updated. Added error checks and updated to the news gnark release and it really improved performance. Thanks again. It is ready for review.

Therecanbeonlyone1969 reviewed (approved):
All requested changes included. Looks good from my end.

rfisch commented:
I will defer my request to review to someone else more qualified in both ZKP and Go.

myrade commented:
Changes look good.

GoldenBit0 commented:
3 maintainer approvals - merging
